bundles/org.eclipse.equinox.preferences/src/org/eclipse/core/internal/preferences/BundleStateScopeServiceFactory.java

```diff
@@ -0,0 +1,75 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Christoph Läubrich and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.core.internal.preferences;
+
+import org.eclipse.core.internal.runtime.MetaDataKeeper;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.preferences.*;
+import org.eclipse.osgi.service.datalocation.Location;
+import org.osgi.framework.Bundle;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.*;
+
+/**
+ * The Bundle State Scope supply bundles with two items:
+ * <ul>
+ * <li>{@link #getNode(String)} returns a node for the bundle workspace
+ * preferences</li>
+ * <li>{@link #getLocation()} returns the state location of the bundle that
+ * acquires the service</li>
+ * </ul>
+ * The service could be acquired for example with declarative services the
+ * following way:
+ *
+ * <pre>
+ * @Reference(target = "(type=bundle)")
+ * private IScopeContext scopeContext;
+ * </pre>
+ */
+@Component(service = IScopeContext.class, property = "type=bundle", scope = ServiceScope.BUNDLE)
+public class BundleStateScopeServiceFactory implements IScopeContext {
+
+	private Bundle usingBundle;
+	private IPath bundleLocation;
+
+	// this is not used, but makes sure the service is not available before a
+	// workspace url is set
+	@Reference(target = "(&(&(objectClass=org.eclipse.osgi.service.datalocation.Location)(type=osgi.instance.area))(url=*))")
+	Location workspaceLocation;
+
+	@Activate
+	void activate(ComponentContext context) {
+		usingBundle = context.getUsingBundle();
+	}
+
+	@Override
+	public String getName() {
+		return "bundle_state"; //$NON-NLS-1$
+	}
+
+	@Override
+	public IEclipsePreferences getNode(String qualifier) {
+		return (IEclipsePreferences) InstanceScope.INSTANCE.getNode(usingBundle.getSymbolicName()).node(qualifier);
```

Comment:
On deactivate is there anything that could be cleaned up?  Like the listeners that may have been added?  Or are there cases where one bundle listens to another bundles symbolic named node?  I've never been that familiar with how all this IEclipsePreferences stuff is used.

Reply:
I guess there is no reason to cache the `getNode(usingBundle.getSymbolicName()` call here?  Perhaps assign to a final in the constructor?